torch/_inductor/codegen/triton.py

```diff
@@ -722,65 +723,56 @@ def indexing(
+                # indirect indexing
+                cse_var = self.cse.varname_map[var.name]
+                mask_vars.update(cse_var.mask_vars)
+            else:
+                # var is one of xN, yN or rN
+                assert var.name[0] in "xyr", var.name
+                mask_vars.add(f"{var.name[0]}mask")
 
         need_dense = (
             config.triton.dense_indexing
             or dense_indexing
-            or indirect_indexing
             or self._load_mask is not None
         ) and index != 0
 
         have_dense = True
         have_loop_vars = False
-        mask = []
-        dense_mask = []
+        dense_mask = set()
 
         for tree in self.range_trees:
             if tree.prefix == "r" and not self.inside_reduction:
                 continue
             if index_vars.intersection(tree.var_list):
                 have_loop_vars = True
                 have_dense = False
-                mask.append(f"{tree.prefix}mask")
-            dense_mask.append(f"{tree.prefix}mask")
+            dense_mask.add(f"{tree.prefix}mask")
 
         if (need_dense and not have_dense) or isinstance(index, sympy.Integer):
             index_str = f"{index_str} + tl.zeros({self.dense_size_str()}, tl.int32)"
             if isinstance(index, sympy.Integer):
-                return index_str, "None"
+                return index_str, set(), "None"
             else:
-                mask = dense_mask
-
+                mask_vars = dense_mask
         elif not have_loop_vars and copy_shape:
-            mask = dense_mask
+            mask_vars = dense_mask
             index_str = f"{index_str} + tl.zeros({copy_shape}.shape, tl.int32)"
-        elif indirect_indexing:
-            # Use dense mask for indirect_indexing
-            # See https://github.com/pytorch/torchdynamo/issues/1654
-            # TODO - An optimization could be to hoist this load outside of
-            # reduction loop, if it is independent of rmask. Such example can be found in
-            # https://github.com/pytorch/torchdynamo/issues/1654
-            index_str = f"{index_str} + tl.zeros({self.dense_size_str()}, tl.int32)"
-            mask = dense_mask
 
         if self._load_mask:
-            mask.append(self._load_mask)
-        elif not mask:
-            mask = ["None"]
+            mask_vars.add(self._load_mask)
 
-        if mask == ["xmask"] and index == 0 and self.range_trees[0].numel == 1:
+        if mask_vars == {"xmask"} and index == 0 and self.range_trees[0].numel == 1:
             # This causes a triton error:
             # https://github.com/openai/triton/issues/633
-            mask = ["None"]
+            mask_vars = set()
 
-        if (
-            index_str in self.cse.varname_map
-            and self.cse.varname_map[index_str].is_scalar
-        ):
-            mask = ["None"]
-
-        return index_str, " & ".join(map(str, mask))
+        mask_str = " & ".join(map(str, mask_vars)) if mask_vars else "None"
```

Comment:
```suggestion
        mask_str = " & ".join(map(str, sorted(mask_vars))) if mask_vars else "None"
```
Not sure if there are other places this will be needed too.

Reply:
Thanks, I believe this is the only one.